// ngx-barcode6/esm2020/lib/ngx-barcode6.module.mjs
import { NgModule } from '@angular/core';
import { NgxBarcode6Component } from './ngx-barcode6.component';
import * as i0 from "@angular/core";
export class NgxBarcode6Module {
}
NgxBarcode6Module.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.1", ngImport: i0, type: NgxBarcode6Module, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgxBarcode6Module.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.0.1", ngImport: i0, type: NgxBarcode6Module, declarations: [NgxBarcode6Component], exports: [NgxBarcode6Component] });
NgxBarcode6Module.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.0.1", ngImport: i0, type: NgxBarcode6Module });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.1", ngImport: i0, type: NgxBarcode6Module, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        NgxBarcode6Component
                    ],
                    imports: [],
                    exports: [
                        NgxBarcode6Component
                    ]
                }]
        }] });
