// @manekinekko/angular-web-bluetooth/esm2020/lib/test.utils.mjs
/*
 * Fake Web Bluetooth implementation
 * Replace real browser Bluetooth objects by a much simpler objects that implement some required functionalities
 */
export class FakeBluetoothDevice {
    constructor(id, name) {
        this.id = id;
        this.name = name;
        this.listeners = {
            gattserverdisconnected: []
        };
    }
    addEventListener(type, listener) {
        this.listeners[type] = [
            ...this.listeners[type],
            listener
        ];
    }
    disconnect() {
        const mockedEvent = { target: this };
        this.listeners.gattserverdisconnected.forEach(listener => listener(mockedEvent));
    }
    clear() {
        this.id = undefined;
        this.name = undefined;
        this.listeners = {
            gattserverdisconnected: []
        };
    }
}
export class FakeBluetoothRemoteGATTServer {
    constructor(device, services) {
        this.device = device;
        this.services = services;
        this.connected = false;
        device.gatt = this;
    }
    connect() {
        this.connected = true;
        return Promise.resolve(this);
    }
    getPrimaryService(service) {
        return Promise.resolve(this.services[service].service);
    }
    disconnect() {
        this.device.disconnect();
        this.connected = false;
    }
}
export class FakeBluetoothRemoteGATTService {
    constructor(device, characteristics) {
        this.device = device;
        this.characteristics = characteristics;
        this.characteristics.service = this;
    }
    getCharacteristic(characteristic) {
        return Promise.resolve(this.characteristics[characteristic]);
    }
}
export class FakeBluetoothRemoteGATTCharacteristic {
    constructor(properties, initialValue) {
        this.listeners = {
            characteristicvaluechanged: []
        };
        this.properties = properties;
        this.value = initialValue;
        this.initialValue = initialValue;
    }
    readValue() {
        return Promise.resolve(this.value);
    }
    addEventListener(type, listener) {
        this.listeners[type] = [
            ...this.listeners[type],
            listener
        ];
    }
    changeValue(value) {
        this.value = value;
        const mockedEvent = { target: this };
        this.listeners.characteristicvaluechanged.forEach(listener => listener(mockedEvent));
    }
    clear() {
        this.value = this.initialValue;
        this.listeners = {
            characteristicvaluechanged: []
        };
    }
}
