// nexti-lib-datatable/esm2015/lib/component/titlesecondary/titlesecondary.component.js
import { Component, Input } from '@angular/core';
export class TitlesecondaryComponent {
    constructor() {
        this.infoAdicional = [];
        this.title = '';
        this.subtitle = '';
    }
    ngOnChanges(changes) {
        var _a, _b;
        const { infoAdicional } = changes;
        if ((infoAdicional === null || infoAdicional === void 0 ? void 0 : infoAdicional.currentValue) !== (infoAdicional === null || infoAdicional === void 0 ? void 0 : infoAdicional.previousValue)) {
            this.title = (_a = this.infoAdicional[0]) === null || _a === void 0 ? void 0 : _a.title;
            this.subtitle = (_b = this.infoAdicional[0]) === null || _b === void 0 ? void 0 : _b.subtitle;
        }
    }
    ngOnInit() {
        var _a, _b;
        this.title = (_a = this.infoAdicional[0]) === null || _a === void 0 ? void 0 : _a.title;
        this.subtitle = (_b = this.infoAdicional[0]) === null || _b === void 0 ? void 0 : _b.subtitle;
    }
}
TitlesecondaryComponent.decorators = [
    { type: Component, args: [{
                selector: 'lib-titlesecondary',
                template: "<h5 class=\"nombre-banco\">{{ title }}</h5>\r\n<p class=\"informacion\">\r\n  {{ subtitle }}\r\n</p>\r\n",
                styles: [".nombre-banco{font-weight:700;font-size:16px;line-height:20px}.informacion,.nombre-banco{font-style:normal;display:flex;align-items:center;color:#0f1b32}.informacion{font-weight:400;font-size:12px;line-height:15px}"]
            },] }
];
TitlesecondaryComponent.ctorParameters = () => [];
TitlesecondaryComponent.propDecorators = {
    infoAdicional: [{ type: Input }]
};
