// nexti-lib-datatable/esm2015/lib/component/table/table.component.js
import { CurrencyPipe } from '@angular/common';
import { Component, Input, Output, EventEmitter, } from '@angular/core';
var NumberFormat;
(function (NumberFormat) {
    NumberFormat["NUM"] = "number";
    NumberFormat["CUR"] = "currency";
})(NumberFormat || (NumberFormat = {}));
var FormatType;
(function (FormatType) {
    FormatType["NUM"] = "number";
    FormatType["CUR"] = "currency";
    FormatType["COU"] = "count";
})(FormatType || (FormatType = {}));
var OperationFormat;
(function (OperationFormat) {
    OperationFormat["SUM"] = "suma";
    OperationFormat["RES"] = "resta";
})(OperationFormat || (OperationFormat = {}));
export class TableComponent {
    constructor(currencyPipe) {
        this.currencyPipe = currencyPipe;
        this.paginacion = false;
        this.totalRegistros = 0;
        this.actualPage = 0;
        this.multFPP = 0;
        this.onNext = new EventEmitter();
        this.onReturn = new EventEmitter();
        this.onDelete = new EventEmitter();
        this.onUpdate = new EventEmitter();
        this.onActive = new EventEmitter();
        this.onShow = new EventEmitter();
        this.onShowEvent = new EventEmitter();
        this.footerRow = [];
        this.dataPerm$ = new Array();
        this.observableBtnDelete$ = false;
        this.observableBtnUpdate$ = false;
        this.observableBtnActive = false;
        this.filas = [];
        this.filasPorPagina = 0;
    }
    ngOnChanges(changes) {
        const { data$, recursos } = changes;
        if ((data$ === null || data$ === void 0 ? void 0 : data$.currentValue) !== (data$ === null || data$ === void 0 ? void 0 : data$.previousValue)) {
            this.dataPerm$ = data$ === null || data$ === void 0 ? void 0 : data$.currentValue;
        }
        if ((recursos === null || recursos === void 0 ? void 0 : recursos.currentValue) !== (recursos === null || recursos === void 0 ? void 0 : recursos.previousValue)) {
            this.observableBtnUpdate$ = this.recursos.update;
            this.observableBtnDelete$ = this.recursos.delete;
            this.observableBtnActive = this.recursos.active;
        }
    }
    ngOnInit() {
        this.orderHeaderStatus = this.header.map((h) => false);
        if (this.paginacion && this.multFPP != 0) {
            this.calcFilasPorPagina();
        }
        else if (this.paginacion && this.multFPP === 0) {
            this.filasPorPagina = 10;
            this.filas.push(10);
        }
    }
    calcFilasPorPagina() {
        let cantidadFilas = 0;
        for (let index = 0; index < 10; index++) {
            cantidadFilas = cantidadFilas + this.multFPP;
            this.filas.push(cantidadFilas);
        }
        this.filasPorPagina = this.filas[0];
    }
    paginaSiguiente() {
        this.onNext.emit(this.filasPorPagina);
    }
    paginaAnterior() {
        this.onReturn.emit(this.filasPorPagina);
    }
    onOrderColumn(head, index) {
        const orderStatus = this.orderHeaderStatus[index];
        this.orderHeaderStatus[index] = !orderStatus;
        this.data$ = this.dataPerm$.sort((_, b) => {
            if (typeof _[index] === 'object') {
                if (_[index].title) {
                    if (_[index].title > b[index].title) {
                        return orderStatus ? 1 : -1;
                    }
                    if (_[index].title < b[index].title) {
                        return orderStatus ? -1 : 1;
                    }
                    return 0;
                }
                if (b[index].subtitle) {
                    if (_[index].subtitle > b[index].subtitle) {
                        return orderStatus ? 1 : -1;
                    }
                    if (_[index].subtitle < b[index].subtitle) {
                        return orderStatus ? -1 : 1;
                    }
                    return 0;
                }
            }
            else {
                if (_[index] > b[index]) {
                    return orderStatus ? 1 : -1;
                }
                if (_[index] < b[index]) {
                    return orderStatus ? -1 : 1;
                }
                return 0;
            }
        });
    }
    getFooterData() {
        var _a;
        const { customTotalTitle, customTotalRows, customTotalFormatting, customTotalType, customTotalAlign, customTotalFixed, customOperationFormatting, } = this.footerConfig;
        let tableFooterData = [];
        let rawData = [];
        (_a = this.data$) === null || _a === void 0 ? void 0 : _a.forEach((itemArray) => {
            let rawDataRow = [];
            // Step 1: format - prepare dat
            const formatedData = itemArray.map((item) => this.getFormatDataToTreatment(item));
            // Step 2: Get raw data (values to totals)
            this.getRawData(customTotalRows, formatedData, customTotalType, rawDataRow);
            rawData.push(rawDataRow);
        });
        if (customTotalRows && customTotalRows.length && rawData.length) {
            let procesedTotals = [];
            switch (customOperationFormatting) {
                case OperationFormat.RES:
                    procesedTotals = this.trasposeMatrix(rawData).map((itemArray) => {
                        return itemArray.reduce((a, b) => a - b);
                    });
                    break;
                default:
                    procesedTotals = this.trasposeMatrix(rawData).map((itemArray) => itemArray.reduce((a, b) => a + b, 0));
                    break;
            }
            if (customTotalFormatting && customTotalFormatting.length) {
                procesedTotals = customTotalFormatting.map((format, index) => {
                    let value, align, fixed, less;
                    switch (format) {
                        case NumberFormat.CUR:
                            value = this.currencyPipe.transform(procesedTotals[index]);
                            break;
                        default:
                            value = procesedTotals[index];
                            break;
                    }
                    align = customTotalAlign[index];
                    fixed = customTotalFixed[index];
                    less = procesedTotals[index] <= -1;
                    return {
                        value,
                        align,
                        fixed,
                        less,
                    };
                });
            }
            tableFooterData = procesedTotals.length ? procesedTotals : null;
        }
        let finalFooterData = [];
        // Clear data
        if (customTotalRows && customTotalRows.length) {
            let cont = 0;
            for (let i = 0; i < this.header.length; i++) {
                let itemToPush;
                customTotalRows.forEach((customIndex) => {
                    if (customIndex === i) {
                        itemToPush = tableFooterData[cont];
                        cont++;
                    }
                });
                if (i === 0 && customTotalTitle) {
                    itemToPush = {
                        fixed: 'td-sticky-left title-total',
                        value: customTotalTitle,
                    };
                }
                if (this.header.length === i + 1 && !this.hiddenActions) {
                    itemToPush = { fixed: 'td-sticky-right' };
                }
                finalFooterData.push(itemToPush);
            }
        }
        this.footerRow = finalFooterData;
    }
    formatCurrencyToNumber(value) {
        if (value !== undefined && value !== null)
            return Number(String(value).replace(/[^0-9.-]+/g, ''));
    }
    trasposeMatrix(matrix) {
        return matrix[0].map((_, colIndex) => matrix.map((row) => row[colIndex]));
    }
    toggleFilterClass(index) {
        return `bi ${this.orderHeaderStatus[index] ? 'bi-caret-up-fill' : 'bi-caret-down-fill'} icon-order`;
    }
    isObject(val) {
        return typeof val === 'object';
    }
    getRawData(customTotalRows, formatedData, customTotalType, rawDataRow) {
        if (customTotalRows && customTotalRows.length) {
            customTotalRows.forEach((customIndex, indexCol) => {
                formatedData.forEach((item, index) => {
                    if (customIndex === index) {
                        let returnItem = 0;
                        customTotalType.forEach((type, indexType) => {
                            if (indexCol === indexType) {
                                switch (type) {
                                    case FormatType.NUM:
                                        if ((typeof item === 'string' &&
                                            (item == '--' || item === '')) ||
                                            typeof item === 'undefined') {
                                            returnItem = 0;
                                        }
                                        else {
                                            returnItem = Number(item);
                                        }
                                        break;
                                    case FormatType.CUR:
                                        if ((typeof item === 'string' &&
                                            (item == '--' || item === '')) ||
                                            typeof item === 'undefined') {
                                            returnItem = 0;
                                        }
                                        else {
                                            returnItem = this.formatCurrencyToNumber(item);
                                        }
                                        break;
                                    case FormatType.COU:
                                        if ((typeof item === 'string' &&
                                            (item == '--' || item === '')) ||
                                            typeof item === 'undefined') {
                                            returnItem = 0;
                                        }
                                        else {
                                            returnItem = 1;
                                        }
                                        break;
                                    default:
                                        returnItem = 0;
                                }
                            }
                        });
                        rawDataRow.push(returnItem);
                    }
                });
            });
        }
    }
    getFormatDataToTreatment(item) {
        if (typeof item === 'object') {
            if (item != null) {
                if (item.title && item.subtitle) {
                    return `${item.title}\n${item.subtitle}`;
                }
                else {
                    return item.title ? item.title : item.subtitle && item.subtitle;
                }
            }
            else {
                return '';
            }
        }
        else {
            return item;
        }
    }
    onDeleteItem(item) {
        this.onDelete.emit(item);
    }
    onActiveItem(item) {
        this.onActive.emit(item);
    }
    onUpdateItem(item) {
        this.onUpdate.emit(item);
    }
    onShowItem(item, $event) {
        this.onShow.emit(item);
        this.onShowEvent.emit($event);
    }
}
TableComponent.decorators = [
    { type: Component, args: [{
                selector: 'nexti-lib-table',
                template: "<div class=\"content-table\">\r\n  <div class=\"content-body table-responsive\">\r\n    <table [attr.aria-label]=\"title\" aria class=\"table\" id=\"table\">\r\n      <thead class=\"table-head\">\r\n        <tr>\r\n          <ng-container *ngFor=\"let head of header; index as i\">\r\n            <ng-container *ngIf=\"isObject(head); else thsimple\">\r\n              <ng-container *ngIf=\"head?.fixedLeft === true\">\r\n                <th\r\n                  scope=\"col\"\r\n                  [ngClass]=\"{\r\n                    'th-sticky-left': true,\r\n                    alignEnd: head?.alignEnd\r\n                  }\"\r\n                >\r\n                  <div class=\"thflex\">\r\n                    <span>{{ head.title }}</span>\r\n                    <ng-container *ngIf=\"head?.order === true\">\r\n                      <i\r\n                        aria-hidden=\"true\"\r\n                        [class]=\"toggleFilterClass(i)\"\r\n                        (click)=\"onOrderColumn(head, i)\"\r\n                      ></i>\r\n                    </ng-container>\r\n                  </div>\r\n                </th>\r\n              </ng-container>\r\n              <ng-container *ngIf=\"head?.fixedRight === true\">\r\n                <th\r\n                  scope=\"col\"\r\n                  [ngClass]=\"{\r\n                    'th-sticky-right': true,\r\n                    alignEnd: head?.alignEnd\r\n                  }\"\r\n                >\r\n                  <div class=\"thflex\">\r\n                    <span>{{ head.title }}</span>\r\n                    <ng-container *ngIf=\"head?.order === true\">\r\n                      <i\r\n                        aria-hidden=\"true\"\r\n                        [class]=\"toggleFilterClass(i)\"\r\n                        (click)=\"onOrderColumn(head, i)\"\r\n                      ></i>\r\n                    </ng-container>\r\n                  </div>\r\n                </th>\r\n              </ng-container>\r\n              <ng-container *ngIf=\"!head?.fixedRight && !head?.fixedLeft\">\r\n                <th scope=\"col\">\r\n                  <div\r\n                    class=\"thflex\"\r\n                    [ngClass]=\"{\r\n                      alignEnd: head?.alignEnd\r\n                    }\"\r\n                  >\r\n                    <span>{{ head.title }}</span>\r\n                    <ng-container *ngIf=\"head?.order === true\">\r\n                      <i\r\n                        aria-hidden=\"true\"\r\n                        [class]=\"toggleFilterClass(i)\"\r\n                        (click)=\"onOrderColumn(head, i)\"\r\n                      ></i>\r\n                    </ng-container>\r\n                  </div>\r\n                </th>\r\n              </ng-container>\r\n            </ng-container>\r\n            <ng-template #thsimple>\r\n              <th scope=\"col\">\r\n                <span>{{ head }}</span>\r\n              </th>\r\n            </ng-template>\r\n          </ng-container>\r\n        </tr>\r\n      </thead>\r\n      <ng-container *ngIf=\"data$ && data$.length > 0\">\r\n        <tbody class=\"table-body\">\r\n          <tr *ngFor=\"let list of data$\">\r\n            <ng-container *ngFor=\"let item of list\">\r\n              <ng-container *ngIf=\"!item?.hidden\">\r\n                <ng-container *ngIf=\"isObject(item); else tdsimple\">\r\n                  <ng-container\r\n                    *ngIf=\"item?.fixedLeft || item?.fixedRight; else tdsimple\"\r\n                  >\r\n                    <ng-container\r\n                      *ngIf=\"item?.fixedLeft === true; else tdfixedright\"\r\n                    >\r\n                      <td\r\n                        [ngClass]=\"{\r\n                          'td-sticky-left': true,\r\n                          alignEnd: item?.alignEnd ? true : false\r\n                        }\"\r\n                      >\r\n                        <ng-container *ngIf=\"item?.title\">\r\n                          <a\r\n                            href=\"#\"\r\n                            *ngIf=\"item.link\"\r\n                            (click)=\"onShowItem(item, $event)\"\r\n                            data-bs-toggle=\"modal\"\r\n                            data-bs-target=\"#exampleModal\"\r\n                            [attr.data-id]=\"item?.title\"\r\n                            >{{ item?.title }}</a\r\n                          >\r\n                          <span\r\n                            [ngClass]=\"{\r\n                              title: item.link,\r\n                              warn: formatCurrencyToNumber(item?.title) <= -1\r\n                            }\"\r\n                            *ngIf=\"!item?.link\"\r\n                            >{{ item?.title }}</span\r\n                          >\r\n                        </ng-container>\r\n                        <ng-container *ngIf=\"item?.title && item?.subtitle\">\r\n                          <br />\r\n                        </ng-container>\r\n                        <ng-container *ngIf=\"item?.subtitle\">\r\n                          <span\r\n                            class=\"subtitle\"\r\n                            [ngClass]=\"{\r\n                              warn: formatCurrencyToNumber(item?.subtitle) <= -1\r\n                            }\"\r\n                            >{{ item?.subtitle }}</span\r\n                          >\r\n                        </ng-container>\r\n                      </td>\r\n                    </ng-container>\r\n                    <ng-template #tdfixedright>\r\n                      <td\r\n                        [ngClass]=\"{\r\n                          'td-sticky-right': true,\r\n                          alignEnd: item?.alignEnd ? true : false\r\n                        }\"\r\n                      >\r\n                        <ng-container *ngIf=\"item?.title\">\r\n                          <a\r\n                            href=\"#\"\r\n                            *ngIf=\"item.link\"\r\n                            (click)=\"onShowItem(item, $event)\"\r\n                            data-bs-toggle=\"modal\"\r\n                            data-bs-target=\"#exampleModal\"\r\n                            [attr.data-id]=\"item?.title\"\r\n                            >{{ item?.title }}</a\r\n                          >\r\n                          <span\r\n                            [ngClass]=\"{\r\n                              title: item.link,\r\n                              warn: formatCurrencyToNumber(item?.title) <= -1\r\n                            }\"\r\n                            *ngIf=\"!item?.link\"\r\n                            >{{ item?.title }}</span\r\n                          >\r\n                        </ng-container>\r\n                        <ng-container *ngIf=\"item?.title && item?.subtitle\">\r\n                          <br />\r\n                        </ng-container>\r\n                        <ng-container *ngIf=\"item?.subtitle\">\r\n                          <span\r\n                            class=\"subtitle\"\r\n                            [ngClass]=\"{\r\n                              warn: formatCurrencyToNumber(item?.subtitle) <= -1\r\n                            }\"\r\n                            >{{ item?.subtitle }}</span\r\n                          >\r\n                        </ng-container>\r\n                      </td>\r\n                    </ng-template>\r\n                  </ng-container>\r\n                  <ng-template #tdsimple>\r\n                    <td\r\n                      [ngClass]=\"{\r\n                        alignEnd: item?.alignEnd ? true : false\r\n                      }\"\r\n                    >\r\n                      <a\r\n                        href=\"#\"\r\n                        *ngIf=\"item?.link\"\r\n                        (click)=\"onShowItem(item, $event)\"\r\n                        data-bs-toggle=\"modal\"\r\n                        data-bs-target=\"#exampleModal\"\r\n                        [attr.data-id]=\"item?.title\"\r\n                        >{{ item?.title }}</a\r\n                      >\r\n                      <span\r\n                        [ngClass]=\"{\r\n                          title: item.link,\r\n                          warn: formatCurrencyToNumber(item?.title) <= -1\r\n                        }\"\r\n                        *ngIf=\"!item?.link\"\r\n                        >{{ item?.title }}</span\r\n                      >\r\n                      <br /><span\r\n                        class=\"subtitle\"\r\n                        [ngClass]=\"{\r\n                          warn: formatCurrencyToNumber(item?.subtitle) <= -1\r\n                        }\"\r\n                        >{{ item?.subtitle }}</span\r\n                      >\r\n                    </td>\r\n                  </ng-template>\r\n                </ng-container>\r\n                <ng-template #tdsimple>\r\n                  <td\r\n                    [ngClass]=\"{\r\n                      warn: formatCurrencyToNumber(item) <= -1\r\n                    }\"\r\n                  >\r\n                    {{ item }}\r\n                  </td>\r\n                </ng-template>\r\n              </ng-container>\r\n            </ng-container>\r\n            <ng-container *ngIf=\"!hiddenActions\">\r\n              <td class=\"td-sticky-right\">\r\n                <ng-container *ngIf=\"observableBtnUpdate$ && !actionState\">\r\n                  <a\r\n                    class=\"icon-actions\"\r\n                    (click)=\"onUpdateItem(list)\"\r\n                    data-bs-toggle=\"modal\"\r\n                    data-bs-target=\"#exampleModal\"\r\n                    ><i\r\n                      aria-hidden=\"true\"\r\n                      class=\"bi bi-pencil-square secondary\"\r\n                    ></i\r\n                  ></a>\r\n                </ng-container>\r\n                <ng-container *ngIf=\"observableBtnDelete$ && !actionState\">\r\n                  <a class=\"icon-actions\" (click)=\"onDeleteItem(list)\"\r\n                    ><i aria-hidden=\"true\" class=\"bi bi-trash primary\"></i\r\n                  ></a>\r\n                </ng-container>\r\n                <ng-container *ngIf=\"observableBtnActive && actionState\">\r\n                  <button\r\n                    (click)=\"onActiveItem(list)\"\r\n                    type=\"button\"\r\n                    class=\"btn btn-primary\"\r\n                  >\r\n                    Activar\r\n                  </button>\r\n                </ng-container>\r\n              </td>\r\n            </ng-container>\r\n          </tr>\r\n        </tbody>\r\n        <tfoot *ngIf=\"footerConfig\">\r\n          <tr>\r\n            <ng-container *ngFor=\"let item of footerRow\">\r\n              <td\r\n                [classList]=\"item?.fixed\"\r\n                [ngClass]=\"{ warn: item?.less }\"\r\n                [style]=\"item?.align\"\r\n              >\r\n                {{ item?.value }}\r\n              </td>\r\n            </ng-container>\r\n          </tr>\r\n        </tfoot>\r\n      </ng-container>\r\n    </table>\r\n    <div class=\"emptyContent\" *ngIf=\"!data$ || data$.length === 0\">\r\n      <ng-container *ngIf=\"!isLoading\">\r\n        <img class=\"icon-empty\" src=\"./assets/icons/empty.svg\" alt=\"empty\" />\r\n        <h6>{{ titleEmpty }}</h6>\r\n      </ng-container>\r\n    </div>\r\n  </div>\r\n  <div\r\n    class=\"content-footer d-flex justify-content-end pt-3\"\r\n    *ngIf=\"!opOcultarFooter\"\r\n  >\r\n    <ng-container *ngIf=\"data$?.length !== undefined\">\r\n      <ng-container *ngIf=\"paginacion; else comunfooter\">\r\n        <div class=\"d-flex w-50 justify-content-end align-items-start\">\r\n          <div class=\"d-flex flex-row align-items-center\">\r\n            <h6 class=\"text-size-sm muted mx-2 mb-0\">Filas por p\u00E1gina:</h6>\r\n            <select\r\n              class=\"form-select custom-select\"\r\n              [(ngModel)]=\"filasPorPagina\"\r\n            >\r\n              <option *ngFor=\"let item of filas\" value=\"{{ item }}\">\r\n                {{ item }}\r\n              </option>\r\n            </select>\r\n            <h6 class=\"text-size-sm muted mx-2 mb-0 text-end\">\r\n              {{\r\n                data$.length ? data$.length * actualPage - data$.length + 1 : 0\r\n              }}\r\n              - {{ actualPage * data$?.length }} de\r\n              {{ totalRegistros }}\r\n            </h6>\r\n            <div class=\"\">\r\n              <i\r\n                class=\"bi bi-chevron-left mx-2 custom-chevron\"\r\n                aria-hidden=\"true\"\r\n                (click)=\"paginaAnterior()\"\r\n              ></i>\r\n              <i\r\n                class=\"bi bi-chevron-right custom-chevron\"\r\n                aria-hidden=\"true\"\r\n                (click)=\"paginaSiguiente()\"\r\n              ></i>\r\n            </div>\r\n          </div>\r\n        </div>\r\n      </ng-container>\r\n      <ng-template #comunfooter>\r\n        <h6 class=\"text-size-sm muted\">\r\n          Mostrando {{ data$?.length }} registros\r\n        </h6>\r\n      </ng-template>\r\n    </ng-container>\r\n  </div>\r\n</div>\r\n",
                styles: ["@import \"https://cdn.jsdelivr.net/npm/bootstrap-icons@1.5.0/font/bootstrap-icons.css\";tfoot{position:sticky;bottom:0;z-index:2}tfoot td{border-collapse:initial;border-top:2px solid #0f1b32!important}.thflex{display:flex;flex-direction:row;align-items:center;justify-content:space-between}.alignEnd{justify-content:flex-end;text-align:end}.table-responsive{width:100%;overflow:auto}.table{width:100%;border-spacing:0}.table,.table-head{border-collapse:initial}.table-head{position:sticky;top:0;z-index:2}.table-head th{border-bottom:2px solid #0f1b32!important}.table-body,.table-body td,.table-head th{border-collapse:initial}.table-body td{border-bottom:1px solid #e4e9ef!important}.primary{color:#0f1b32}.secondary{color:#2a8bf2}.icon-order{cursor:pointer;padding-left:10px}th{font-weight:700;color:#0f1b32;vertical-align:middle}td,th{font-size:12px;line-height:17px;width:auto;height:50px;min-width:150px;max-width:200px;background-color:#fff}td{font-weight:500;color:#585858}.title{color:#2a8bf2}.title-total{font-weight:700}.subtitle{text-transform:uppercase}.icon-actions{font-size:15px;cursor:pointer;color:#585858}.icon-actions:nth-child(2){margin-left:15px}.text-size-sm{font-size:12px;min-width:100px}.th-sticky-left{position:sticky;left:0;z-index:3}.th-sticky-right{position:sticky;right:0;z-index:3}.td-sticky-left{position:sticky;left:0;z-index:1}.td-sticky-right{position:sticky;right:0;z-index:1}.emptyContent{position:sticky;left:0;right:0;width:100%;min-height:200px;display:flex;flex-direction:column;justify-content:center;align-items:center}h6{color:#0f1b32;font-weight:700!important}.icon-empty{font-size:75px}.warn{color:#eb5757!important}.content-table{display:grid;width:100%;height:100%;grid-template-rows:auto 1fr;grid-template-areas:\"content-body\" \"content-footer\"}.content-body{grid-area:content-body}.content-footer{grid-area:content-footer}.custom-select{padding:5px;width:60px;max-width:60px;min-width:60px}.custom-chevron{cursor:pointer}"]
            },] }
];
TableComponent.ctorParameters = () => [
    { type: CurrencyPipe }
];
TableComponent.propDecorators = {
    title: [{ type: Input }],
    titleEmpty: [{ type: Input }],
    header: [{ type: Input }],
    data$: [{ type: Input }],
    footerConfig: [{ type: Input }],
    recursos: [{ type: Input }],
    actionState: [{ type: Input }],
    opOcultarFooter: [{ type: Input }],
    hiddenActions: [{ type: Input }],
    isLoading: [{ type: Input }],
    paginacion: [{ type: Input }],
    totalRegistros: [{ type: Input }],
    actualPage: [{ type: Input }],
    multFPP: [{ type: Input }],
    onNext: [{ type: Output }],
    onReturn: [{ type: Output }],
    onDelete: [{ type: Output }],
    onUpdate: [{ type: Output }],
    onActive: [{ type: Output }],
    onShow: [{ type: Output }],
    onShowEvent: [{ type: Output }]
};
