// nexti-lib-datatable/esm2015/lib/component/breadcrumbs/breadcrumbs.component.js
import { Component, Input } from '@angular/core';
export class BreadcrumbsComponent {
    constructor() {
        this.breadcrumbs = [];
    }
    ngOnInit() {
    }
}
BreadcrumbsComponent.decorators = [
    { type: Component, args: [{
                selector: 'lib-breadcrumbs',
                template: "\r\n<ul class=\"navega\">\r\n    <li *ngFor=\"let item of breadcrumbs\">\r\n        <a *ngIf=\"item.link\" href=\"{{item.link}}\">{{item.title}}</a>\r\n        <span *ngIf=\"!item.link\">{{item.title}}</span>\r\n    </li>\r\n</ul>\r\n",
                styles: [".breadcrumbs{margin:0;padding:0;font-size:12px;line-height:15px;color:#0f1b32}.breadcrumbs span{color:#b0b0b0}.navega{padding:0;margin:0;display:flex;list-style:none}.navega li{margin:0;padding-right:3px;font-size:12px;line-height:15px}.navega a{color:#0f1b32;text-decoration:none}.navega span{color:#b0b0b0}"]
            },] }
];
BreadcrumbsComponent.ctorParameters = () => [];
BreadcrumbsComponent.propDecorators = {
    breadcrumbs: [{ type: Input }]
};
